### PR TITLE
Make macro scope a real name scope and fix some details

### LINKED ISSUE #1688: Most of macros from std are not resolved
I'm using rust-analyzer from [master](https://github.com/rust-analyzer/rust-analyzer/commit/b1b6a90ab168c3d04038a28a3c02c69a507a3f52) with vscode extension.

I found it failed to infer the type of `vec!` (no type decorator, even no `{unknown}`), but it works well with my simple implementation using `macro_rules`. So I think there is something wrong  resolving them.

Most of macros from std do not work, except `std::dbg!`, while `dbg!` (without prefix) still fails.

The output from "Rust Analyzer Language Server Trace" contains no errors.

Screenshot:
Screenshot_20190816_020553

### PR DESCRIPTION
**Pull request by uHOOCCOOHu**
This PR make macro's module scope a real name scope in `PerNs`, instead of handling `Either<PerNs, MacroDef>` everywhere.

In `rustc`, the macro scope behave exactly the same as type and value scope.
It is valid that macros, types and values having exact the same name, and a `use` statement will import all of them. This happened to module `alloc::vec` and macro `alloc::vec!`.
So `Either` is not suitable here.

There is a trap that not only does `#[macro_use]` import all `#[macro_export] macro_rules`, but also imports all macros `use`d in the crate root.
In other words, it just _imports all macros in the module scope of crate root_. (Visibility of `use` doesn't matter.)

And it also happened to `libstd` which has `use alloc_crate::vec;` in crate root to re-export `alloc::vec`, which it both a module and a macro.
The current implementation of `#[macro_use] extern crate` doesn't work here, so that is why only macros directly from  `libstd` like `dbg!` work, while `vec!` from `liballoc` doesn't.
This PR fixes this.

Another point is that, after some tests, I figure out that _`macro_rules` does NOT define macro in current module scope at all_.
It defines itself in legacy textual scope. And if `#[macro_export]` is given, it also is defined ONLY in module scope of crate root. (Then being `macro_use`d, as mentioned above)
(Well, the nightly [Declarative Macro 2.0](https://github.com/rust-lang/rust/issues/39412) simply always define in current module scope only, just like normal items do. But it is not yet supported by us)

After this PR, in my test, all non-builtin macros are resolved now. (Hover text for documentation is available) So it fixes #1688 . Since compiler builtin macros are marked as `#[rustc_doc_only_macro]` instead of `#[macro_export]`, we can simply tweak the condition to let it resolved, but it may cause expansion error.

Some critical notes are also given in doc-comments.

Screenshot_20190909_223859

### CODE REVIEW

**Comment by matklad** [comment]
>This PR make macro's module scope a real name scope in PerNs, instead of handling Either<PerNs, MacroDef> everywhere.

Hm, interesting! So, the original motivation for handing macros separately was that macros are only needed in the initial expansion phase. After that, the rest of the hir should assume a fully macro expanded code, and, for example, name resolution operations should be forbidden from returning a macro. However, I think that this reasoning is actually flawed: the main thing that tries to resolve paths is name resolution and macro expansion itself! Basically the rest of the hir should see paths already resolved.

File: crates/ra_hir/src/nameres/per_ns.rs [hunk 1]
```diff
@@ -1,78 +1,93 @@
+use crate::MacroDef;
+
 #[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub enum Namespace {
     Types,
     Values,
+    Macro,
```
Comment (matklad) [inline crates/ra_hir/src/nameres/per_ns.rs]:
```suggestion
    // Note that only type inference uses this enum, and it doesn't care about macros.
    // Macro,
```

File: crates/ra_hir/src/nameres/per_ns.rs [hunk 2]
```diff
@@ -1,78 +1,93 @@
...
     pub fn and_then<U>(self, f: impl Fn(T) -> Option<U>) -> PerNs<U> {
-        PerNs { types: self.types.and_then(&f), values: self.values.and_then(&f) }
+        PerNs {
+            types: self.types.and_then(&f),
+            values: self.values.and_then(&f),
+            macros: self.macros,
+        }
     }
 
     pub fn map<U>(self, f: impl Fn(T) -> U) -> PerNs<U> {
```
Comment (matklad) [inline crates/ra_hir/src/nameres/per_ns.rs]: I feel like `and_then` and `map` should transform macros as well. Does this break anything?

Comment (uHOOCCOOHu) [reply crates/ra_hir/src/nameres/per_ns.rs]: They have different types. I don't know how to do it better. `and_then` is not used. `map` is only called with `map(Resolution::Def)`. Actually, there are only two types, `PerNs<Resolution>` and `PerNs<ModuleDef>`.

Comment (matklad) [reply crates/ra_hir/src/nameres/per_ns.rs]: Ah, righ, `PerNs` is not generic on macro ns... Yeah, this sort-of makes sense, but at the same time is pretty ugly.... I don't know a better solution though, and existing `Either` stuff is not really pretty anyway, so I am fine with it!

Comment (uHOOCCOOHu) [reply crates/ra_hir/src/nameres/per_ns.rs]: Maybe just rename to `map_type_value()` or something? `and_then()` can be simply removed.

Comment (matklad) [reply crates/ra_hir/src/nameres/per_ns.rs]: Let's remove and_then, and leave `map` as is, for know. I have a feeling that we should remove `.map` as well. Like, we use map to construct `PerNs<Resolution>` in resolve, but this type doesn't make sense

**Pull request merged**